patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch

```diff
@@ -0,0 +1,76 @@
+
+Fix #5189 where entities in slime blocks are unreliable.
+
+Add a general mechanism where the entity knows how many ticks into
+the future that it has to be fully ticked. ActivationRange respects
+that and does not cull any of those ticks to the entity.
+
+diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
+index 896d892237b29eb404398db07264eb6f04786754..2f10a937accd4f791ef8ffe924a77e8ab6f86e9b 100644
+--- a/src/main/java/net/minecraft/world/entity/Entity.java
++++ b/src/main/java/net/minecraft/world/entity/Entity.java
+@@ -690,6 +690,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
+ 
+         this.firstTick = false;
+         this.level.getProfiler().pop();
++        // Paper start   - track when all ticks are required for the entity, e.g. after a piston push
++        if (remainingExactTicks > 0) {
++           remainingExactTicks -= 1;
++        }
++        // Paper end
+     }
+ 
+     public void setSharedFlagOnFire(boolean onFire) {
+@@ -3992,4 +3997,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
+         return ((ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
+     }
+     // Paper end
++ 
++    // Paper start   - track when all ticks are required for the entity, e.g. after a piston push
++    public int remainingExactTicks = 0;
++
++    public void requestExactTicks(int ticks) {
++        remainingExactTicks = Math.max(ticks, remainingExactTicks);
++    }
++    // Peter end
+ }
+diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+index 87bedba9ab495edcce289c6665271d92b7165944..3200a75e22d4e935bf324a55f9f0ff4910f33696 100644
+--- a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
++++ b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+@@ -140,6 +140,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
+                                 }
+ 
+                                 entity.setDeltaMovement(e, g, h);
++                                entity.requestExactTicks(10);    // Paper - track when all ticks are required for the entity, e.g. after a piston push
+                                 break;
+                             }
+                         }
+@@ -173,6 +174,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
+     private static void moveEntityByPiston(Direction direction, Entity entity, double d, Direction direction2) {
+         NOCLIP.set(direction);
+         entity.move(MoverType.PISTON, new Vec3(d * (double)direction2.getStepX(), d * (double)direction2.getStepY(), d * (double)direction2.getStepZ()));
++        entity.requestExactTicks(10);    // Paper - track when all ticks are required for the entity, e.g. after a piston push
```

Comment:
Have you tested with lower values?

Reply:
I belive that the moveCollidedEntities is also called for "bounced", and I thought it best to give enough time for the movement to die down. For the stuck entities 2 ticks should be enough. I'll test that, as well if bounced entities are affected by this patch. Back in 10 minutes.

Reply:
If you have some source other places where you pulled the number, feel free to keep it, but add a comment on the line that has said number.

Reply:
After a bit of testing: for items stuck inside a slime block 2 ticks is sufficient to guarantee it starts and stops moving. 
This function is also called for non-stuck items, that are laucnhed ballistically. I think they will move differently depending on if they get all ticks or not while in motion. Only 2 ticks are needed for the ice maker and similar machines based on stuck entities.

On the other hand, I think paper already accepts non-vanilla behaviour for dynamic movement as the 1/4 tick removal exists, so maybe we should lower tihs value to 2? We can always make it configurable in the future if good test cases are found. I suspect that perhaps tnt ender pearl cannons would be impacted by this, but I have no good test case for this.

Reply:
Updated with a chnage to 2 ticks only. This was verfied as sufficient both manually and with the Ilmango deisgn from issue 5189 running over 1200 blocks without issue. Bounced item trajectories are leaft for the future, once someone designs a test case for comparing differences between vanilla and paper.

Reply:
I suppose a test for some ender pearl device could be done to make sure they remain functional? Otherwise, I'm sure the various broken machines will turn up in the issue tracker after merge, anyway.

Reply:
Any machines working before the fix should see no change after the fix, excpet perhaps a tiny bit less variation. It is likely still not vanilla behaviour for bounced entities, but I would need a test case before we delve into that. I suggest we leave that for a future issue...